### PR TITLE
thunderbird: rewrite (68.2.2 -> 68.3.0)

### PR DESCRIPTION
**Pull request by lovesegfault**
###### Motivation for this change
This supersedes #75143, adding Wayland support, bumping to 68.3.0, and also reworking the build a bit.

The reason for reworking is I noticed the Firefox and Thunderbird builds had gone astray, and since they are almost identical, we ought to try and keep them as much in sync as possible.

I would appreciate @andir taking a look at this since he maintains Firefox and probably has some insights into it.

I've also added myself as a maintainer of the package. 

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @edolstra @nbp

### CODE REVIEW

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Do you know when this diff was added? I believe this bug has been fixed.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Even if we can't let `wrapGAppsHook` do the wrapping, could you move this into `preFixup`?
And actual manual wrapping there as well.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: I do not, but I can give it a shot and see if it works; doing that now.

Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Done!

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Fixed.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]:
```suggestion
  makeFlags = lib.optionals enableOfficialBranding [
```

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: does building with waylandSupport cause it to break under x11?

**Review by worldofpeace — CHANGES_REQUESTED** [review]
Still building over here.

**Comment by lovesegfault** [comment]
Before merging I want to take some of the changes of #55026 and integrate them here.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Nope.

**Comment by worldofpeace** [comment]
> Before merging I want to take some of the changes of #55026 and integrate them here.

It'd be cool to also integrate the code style from there, in particular one item per line in arrays.
Not sure about removing the gtk option, but I already prefer having it on default here, maybe we should just remove it.

**Comment by lovesegfault** [comment]
@worldofpeace Done, I left the arrays as-is because I don't like the 1/line look, if I must I will change it.

I simplified the `makeDesktopItem` and `patchelf` stuff, I'd like a suggestion on how to lift the default-toolkit selection off of the script portion, it looks dirty and it's hard to read.

I mean the
```Nix
    "--enable-default-toolkit=cairo-gtk${if gtk3Support then "3${lib.optionalString waylandSupport "-wayland"}" else "2"}"
```
line.

**Comment by worldofpeace** [comment]
> @worldofpeace Done, I left the arrays as-is because I don't like the 1/line look, if I must I will change it.
> 
> I simplified the `makeDesktopItem` and `patchelf` stuff, I'd like a suggestion on how to lift the default-toolkit selection off of the script portion, it looks dirty and it's hard to read.
> 
> I mean the
> 
> ```nix
>     "--enable-default-toolkit=cairo-gtk${if gtk3Support then "3${lib.optionalString waylandSupport "-wayland"}" else "2"}"
> ```
> 
> line.

You could put it in `let in`
```
let 
  toolkitSlug = if gtk3Support then ''
    3${optionalString waylandSupport "-wayland"}
  '' else "2";

  toolkitValue = "cairo-gtk${toolkitSlug}"
in
```

This also reminds me that `waylandSupport` logically implicates `gtk3Support`.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]:
```suggestion
      categories = lib.concatStringsSep ";" [
```

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]:
```suggestion
      mimeType = lib.concatStringsSep ";" [
```

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Fixed

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Fixed.

**Comment by lovesegfault** [comment]
@worldofpeace Not sure why the eval is borked

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]:
```suggestion
    toolkitSlug = if gtk3Support then ''3${lib.optionalString waylandSupport "-wayland"}'' else "2";
```

Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Fixed.

**Comment by vcunat** [comment]
I've been using current state (601b08d0) for a couple hours, and it seems fine.  (But I haven't really looked at the code, so far.)

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (andir) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Can we format these few lines a bit prettier?

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (andir) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: We haven't previously enabled the official branding. I think there was some discussion about enabling it. For Firefox we got explicit permission.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (andir) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: This can probably be removed as soon as we package a Thunderbird >=71.0 since XUL shouldn't be anymore (in use)?

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (andir) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: blank line ;-)

File: pkgs/top-level/all-packages.nix
Comment (andir) [inline pkgs/top-level/all-packages.nix]: any reason for breaking this here? (I am not opposed)

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (andir) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Again, reflowing these here would probably make sense.

**Review by andir — CHANGES_REQUESTED** [review]
I did leave some comments. Mostly about style and due diligence on catching up with our comments in the expression.

I am not a Thunderbird user so I didn't test this.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (vcunat) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: AFAIK the last comments were in [this thread](https://github.com/NixOS/nixpkgs/pull/36449); I'm not sure if that's sufficiently specific to enable it (I haven't seen any reactions to later pings).

Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: I've been talking in the #maildev channel on MozIRC with the maintainers. They see no issues with our current build, but would like me to email the responsible people to get approval per https://www.mozilla.org/en-US/foundation/trademarks/distribution-policy/.

I'm working on this.

**Comment by vcunat** [comment]
> It'd be cool to also integrate the code style from there, in particular one item per line in arrays.

I agree the style does have some advantages, but to me it takes a bit too much space typically and I personally prefer to group them a bit.  BTW, that PR did this only for `*Inputs` and not for the function parameters; I assume you'd want to apply one-item-per-line in there as well, as that's what I see in gnome expressions.

_In any case, I don't care much and I want to avoid spending noticeable amount of time on bikeshedding discussions._

**Comment by lovesegfault** [comment]
How about I just `nixfmt` it and call it a day?

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Applied nixfmt and will accept what the machines say.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Split

File: pkgs/top-level/all-packages.nix
Comment (lovesegfault) [reply pkgs/top-level/all-packages.nix]: To keep it consistent with Firefox, I want to avoid the builds diverging in any nonessential way.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Added FIXME

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: `rjl`on IRC confirms that the option is now gone.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: From IRC:
`i dont know either honestly  .. necko is networking libraries... it might have something to do with captive portal detection`

Gentoo[1] has this enabled for all systems with wifi.

[1]: https://gitweb.gentoo.org/repo/gentoo.git/tree/mail-client/thunderbird/thunderbird-68.3.0.ebuild#n545

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: I think this has been deprecated, at least it's listed with the remaining "old options"

Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: I'll note that the Gentoo ebuild also has no mention of `system-cairo` anymore: https://gitweb.gentoo.org/repo/gentoo.git/tree/mail-client/thunderbird/thunderbird-68.3.0.ebuild

**Comment by lovesegfault** [comment]
Additional changes, to bring it more inline with other distros packages:

1. pulseaudio has been enabled
2. webp and vpx support

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: I've enabled the flag.

**Comment by lovesegfault** [comment]
* enabled necko-wifi

**Comment by lovesegfault** [comment]
* remove `MOZ_LEGACY_PROFILES`
* remove `MOZ_ALLOW_DOWNGRADE`

(on the advice of upstream maintainers)

**Comment by andir** [comment]
>     * remove `MOZ_LEGACY_PROFILES`
> 
>     * remove `MOZ_ALLOW_DOWNGRADE`
> 
> 
> (on the advice of upstream maintainers)

I think that will get us into a funny position. At least for Firefox it treats each new binary location as an incompatible install of Firefox and thus refuses to reuse your existing profiles. That is why those are in there. If you recompile the package with a few bits flipped you should be able to reproduce that.

**Comment by lovesegfault** [comment]
@andir I expected that too, but in my testing so far I have been unable to reproduce that and things work fine. It'd be nice if someone else would confirm no issues, cc. @vcunat

**Comment by worldofpeace** [comment]
> I agree the style does have some advantages, but to me it takes a bit too much space typically and I personally prefer to group them a bit. BTW, that PR did this only for *Inputs and not for the function parameters; I assume you'd want to apply one-item-per-line in there as well, as that's what I see in gnome expressions.

Yeah, this happens to be how we've chosen to go about things in gnome expressions.
And using `nixpkgs-fmt`, I really value diff-ability. Even if the files are long.

**Comment by vcunat** [comment]
I believe I *did* encounter some issues with thunderbird refusing to "downgrade" or something like that, at least when switching testing of -bin and compiled variant (even though my version never decreased).  That was just before we merged some of these variables IIRC (several weeks ago?)

> I really value diff-ability

Yes, but regardless of this I personally want to use tools that nicely highlight smaller changes on longer lines.  (For non-nix code, for example.)

**Comment by lovesegfault** [comment]
I reverted the changes around legacy profiles for the time being, if they complain during the standardization process we can revisit this.

@worldofpeace, @vcunat, @andir: This is ready for a last look and merge.

File: pkgs/applications/networking/mailreaders/thunderbird/default.nix
Comment (worldofpeace) [inline pkgs/applications/networking/mailreaders/thunderbird/default.nix]: I don't think this is even needed, or should be buildInputs (wrapGAppsHook does the work for us)

Comment (lovesegfault) [reply pkgs/applications/networking/mailreaders/thunderbird/default.nix]: Removed :)

**Review by worldofpeace — APPROVED** [review]
Builds and executes. LGTM.

**Comment by worldofpeace** [comment]
@lovesegfault Could you document all the notable changes in the commit msg?

**Comment by lovesegfault** [comment]
@worldofpeace on it :)

**Comment by lovesegfault** [comment]
@worldofpeace done!

**Review by andir — APPROVED** [review]
Looks good. No more complains from my side (for now ;)).

**Comment by andir** [comment]
@worldofpeace once you are happy with functionality/testing/reviews feel free to merge. Nothing against it from my side.

**Comment by worldofpeace** [comment]
Pushed silly nitpicks and the new homepage https://www.thunderbird.net/en-US/

**Comment by worldofpeace** [comment]
Thanks a lot @lovesegfault :stars: 
Was nice helping you, thanks for the commits.

**Comment by lovesegfault** [comment]
Thanks everyone for helping me push this through :)

**Pull request merged**
